server/src/main/java/org/opensearch/common/throttle/AdmissionController.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+package org.opensearch.common.throttle;
+
+import org.opensearch.common.lease.Releasable;
+
+/**
+ * Aims to provide resource based request admission control.
+ * Provides methods for any tracking-object that can be incremented (such as memory size),
+ * Admission control can be applied if configured limit has been reached.
+ */
+public interface AdmissionController {
```

Comment:
Do we intend for plugins to be able to implement this interface? Can we make this package private? 

If not please mark this with an appropriate opensearch (@opensearch.internal, @opensearch.experimental, @opensearch.api) annotation to document extensibility restrictions.

Reply:
> This interface needs to be included in a PR that implements it and includes some kind of tests. We don't merge PRs for unused interfaces. If this is a merge of a fairly large downstream implementation make sure it's split into smaller incremental PRs.
> 
> Also, we'll want to feature flag this so we can let it bake before releasing GA.

Thanks for the feedback @nknize 

I planned to raise implementations as part of next pull-requests to keep scope of each pull-request smaller.

I will update this pull request with other suggested changes.

Reply:
> Do we intend for plugins to be able to implement this interface? Can we make this package private?

Made it package private as it is not intended for plugins to implement this interface.

---

> Also, we'll want to feature flag this so we can let it bake before releasing GA.

Yes, this feature will be behind dynamic setting.